### ISSUE TITLE
[RFC] Kimi-K2.5

### ISSUE THREAD

**Issue by depksingh** [sgl-project/sglang-jax#1011]
#### Summary

This RFC proposes adding inference support for **Kimi-K2.5** (moonshotai/Kimi-K2.5) in sglang-jax's multimodal pipeline. Kimi-K2.5 is a native vision-language model from Moonshot AI that integrates a DeepSeek-V3-style MoE LLM with a custom vision encoder.

##### 1.1 LLM Backbone

The LLM is a **DeepSeek-V3-style** transformer (model\_type: "kimi\_k2") characterized by massive sparsity and memory-efficient attention:

* **Attention:** Uses **Multi-head Latent Attention (MLA)** to compress the KV cache.  
  * **Latent Ranks:** kv\_lora\_rank=512, q\_lora\_rank=1536.  
  * **Head Dimensions:** qk\_nope\_head\_dim=128, qk\_rope\_head\_dim=64, v\_head\_dim=128.  
* **Mixture-of-Experts (MoE):** 61 layers total.  
  * **Layer 0:** Dense (acts as a standardizer).  
  * **Layers 1–60:** Sparse MoE with **384 routed experts** and **1 shared expert**.  
  * **Routing:** Selects **top-8** experts per token using sigmoid scoring with bias correction.  
* **Positional Encoding:** Uses **YaRN RoPE** (factor=64.0, theta=50,000).  
  * **Crucial Note:** Unlike Qwen-VL, Kimi-K2.5 does **not** use MRoPE; vision tokens are treated as standard sequence elements.  
* **Capacity:** 1.04T total parameters (\~32B active), 262,144 token context window, and 163,840 vocabulary size.

##### 1.2 Vision Encoder: MoonViT-3D

Kimi-K2.5 uses the **MoonViT-3D** encoder, a 425M parameter model designed for unified image and video understanding:

* **Structure:** 27 layers, hidden dimension of 1152, and 16 attention heads.  
* **Positional Embedding:** KimiDividedFixedPosEmb (separate H, W, T embeddings).  
* **Spatio-Temporal Attention:** For video, it performs sequential spatial passes (within frames) and temporal passes (across frames) rather than full 3D attention.  
* **Patch Merger:** Reduces token density by pooling 2 times 2 spatial patches and applying temporal pooling before projecting to the LLM's 7168 hidden dimension.

##### 2.1 Component Reuse

The following sglang-jax components are directly compatible:

* [DeepseekV3ForCausalLM](https://github.com/sgl-project/sglang-jax/blob/deepseekv3/python/sgl_jax/srt/models/deepseekv3.py#L392)   
* apply\_linear\_quantization (for bf16 → FP8 on-the-fly conversion).

##### 2.2 New Developments

* **kimi\_vit.py**: A full implementation of MoonViT-3D, including the divided positional embeddings and the 2 times 2 patch merger.  
* **kimi\_k2\_5\_generation.py**: A KimiK25LLMModel subclass that enables the multimodal prefill path (using input\_embedding instead of token IDs for the vision sequence).  
* **Stage Config:** A new YAML defining the Stage 0 (ViT) and Stage 1 (LLM) pipeline topology.

##### 3.1 Kimi Vision Encoder (kimi\_vit.py)

This is the primary development task. The Kimi ViT architecture requires:

1. **KimiDividedFixedPosEmb:** Separate sinusoidal embeddings for height, width, and time, interpolated at runtime.  
2. **KimiVisionAttention:** Implements spatial\_temporal attention (sequential spatial and temporal passes).  
3. **KimiPatchMerger:** A projector that pools 2 times 2 spatial patches and applies temporal pooling, projecting from 4,608 to 7,168 hidden dimensions.

##### 3.2 Multimodal LLM Wrapper

The KimiK25LLMModel (a subclass of DeepseekV3Model) will override the execution call to inspect forward\_batch.input\_embedding. During the **prefill** phase, it will use the merged embeddings from Stage 0 instead of standard token embeddings.

#### 4\. End-to-End Data Flow

1. **Tokenization:** The MultimodalTokenizer processes inputs, replacing media placeholders with media\_placeholder\_token\_id=163605 and skipping MRoPE.  
2. **Stage 0:** The KimiVisionTransformer processes pixel values. The KimiPatchMerger reduces token counts and projects them to the LLM's hidden dimension.  
3. **Embedding Merge:** Vision embeddings are inserted into the text embedding sequence at the placeholder locations.  
4. **Stage 1:** The LLM performs a prefill using the merged embeddings, followed by standard autoregressive decoding for text generation.

<img width="1292" height="1037" alt="Image" src="https://github.com/user-attachments/assets/d2a3a830-cbdb-4618-8d9c-e7c7307d7989" />

#### 5\. Implementation Roadmap

1. **Weight Inspection:** Validate safetensors structure.  
2. **Config Setup:** Register YAML and dataclasses.  
3. **ViT Development:** Implement and isolate-test the Kimi Vision Encoder.  
4. **LLM Integration:** Wrap DeepseekV3 and wire the embedding injection path.  
5. **Runner Logic:** Generalize VitModelRunner to support the Kimi interface.  
6. **Validation:** Compare output logits against the HF reference implementation.

##### 6.1 P0 Milestones

1. Integration with LLM Backbone \[Deepseek v3\]  - 26/05
2. E2E successful run on multihost  - 29/05
3. Accuracy & sanity validations  - 05/06
4. Vision Encoder Implementation \[Image support only\]  - 26/06
5. Video Input support  - 30/06
6. Baselining and comparison with B200 sglang oss version - 30/06
7. Reasoning & Tooling Parser - 30/06
##### 6.2 P1 Milestones

1. Speculative decoding

**Comment by JamesBrianD** [comment]
@pengchengneo Please review the multimodal design.

**Comment by Bob-Chen222** [comment]
Hi @depksingh  could you also list the timeline for P0 milestones for kimik2.5? We want to prioritize this model
And also could you explain a bit on why the Video support is in P1?

**Comment by Bob-Chen222** [comment]
Also please add according unit tests and e2e tests for vision encoder and other modules. Thanks

**Comment by pengchengneo** [comment]
@depksingh  How many resources do you plan to use for this model inference, v6e or v7x? 
If the ViT and LLM backbone involve cross-host data transfer, a DCN solution might be work or you can use pathways to support your MPMD implement(most of our current multimodal inference models are focused on a single host, which support us transfer latent date by ICI net work).

**Comment by depksingh** [comment]
Hey @pengchengneo , sure, we'll take a look into the DCN or pathways, Also, We will be working with v7x for now. 
@Bob-Chen222  added the timelines as discussed.
Thank you.

**Comment by pengchengneo** [comment]
@depksingh 
If multimodal refactoring is involved, we can discuss it. Thank you.

**Comment by depksingh** [comment]
hey @pengchengneo , yes, we'll need some refactoring there, we're currently focusing on the isolated kimi pieces like ViT, VL Generation this week, we can discuss those maybe next week, will share here.

**Comment by cjx0709** [comment]
@depksingh hello, thanks for putting this RFC. Just wanted to check whether there's any update on the implementation side. Is a PR still planned for this proposal? Happy to help review or discuss if there are any blockers.

**Comment by depksingh** [comment]
Hey @cjx0709 , Yes, we are currently working on a fork here: (https://github.com/chaitanyapk/sglang-jax/pulls?q=), We have completed the ViT and vl_generation pieces, but currently working on the E2E pipeline changes and getting it running e2e. The plan was to share the PR wrt sglang main, once we have a fully functional run. But please feel free to pitch in if you have any comments in the existing implementation. And let us know if its fine, if we raise PRs for independent blocks for ViT and vl_generation in the meantime until the e2e run is completed, or we can share the PR once its completed.

**Comment by cjx0709** [comment]
@depksingh Great work! I think you can submit pull requests in stages. For example, you could complete the pull request for the kimi-k2.5 text model and test at first. If the module's functionality is independent and the testing is complete, I think submitting a separate pull request is fine (you could create some sub-issue RFCs for the sub-modules under this current RFC).

**Comment by depksingh** [comment]
Sure, As the model is in INT4 and v7x doesn't support it, @chaitanyapk is working on supporting the INT4 in an emulated way for the Kimi existing changes, once those changes are done, we can share PRs in stages as you mentioned above. Please let us know if you have any concerns.
